File: android/MicroService/src/main/java/org/dweb_browser/microservice/core/BootstrapContext.kt
package org.dweb_browser.microservice.core

import org.dweb_browser.helper.MICRO_MODULE_CATEGORY
import org.dweb_browser.helper.MMID
import org.http4k.core.Request


interface BootstrapContext {
  val dns: DnsMicroModule
}

interface DnsMicroModule {
  /**
   * 动态安装应用
   */
  fun install(mm: MicroModule)

  /**
   * 动态卸载应用
   */
  fun uninstall(mm: MMID): Boolean

  /**
   * 动态js应用查询
   */
  fun query(mmid: MMID): MicroModule?

  /**
   * 根据类目搜索模块
   * > 这里暂时不需要支持复合搜索，未来如果有需要另外开接口
   * @param category
   */
  suspend fun search(category: MICRO_MODULE_CATEGORY): MutableList<MicroModule>

  /**
   * 重启应用
   */
  suspend fun restart(mmid: MMID)

  /**
   * 与其它应用建立连接
   */
  suspend fun connect(mmid: MMID, reason: Request? = null): ConnectResult

  /**
   * 启动其它应用
   */
  suspend fun open(mmid: MMID): Boolean

  /**
   * 关闭其他应用
   */
  suspend fun close(mmid: MMID): Boolean
}

File: desktop-dev/src/core/bootstrapContext.ts
import type { $PromiseMaybe } from "./helper/types.ts";
import type { MicroModule } from "./micro-module.ts";
import type { $ConnectResult } from "./nativeConnect.ts";
import type { $MMID } from "./types.ts";
export interface $BootstrapContext {
  dns: $DnsMicroModule;
}
export interface $DnsMicroModule {
  install(mm: MicroModule): void;
  uninstall(mm: $MMID): Promise<boolean>;
  connect(mmid: $MMID, reason?: Request): $PromiseMaybe<$ConnectResult>;
  query(mmid: $MMID): Promise<MicroModule | undefined>;
  /**
   * 打开应用，如果应用不存在，或者因某种原因（程序错误、或者被限制）启动失败，会返回 false
   * @param mmid
   */
  open(mmid: $MMID): Promise<boolean>;
  /** 关闭应用，如果应用不存在，或者用户拒绝关闭、或者因为某种原因（程序错误、或者被限制），会返回false */
  close(mmid: $MMID): Promise<boolean>;
  restart(mmid: $MMID): void;
}
